<p style="font-family: Arial; font-size:3em;color:black;"> Session 4 - Lab Exercise 7</p>

We will use the titanic dataset from Kaggle (https://www.kaggle.com/). This is a well-known dataset and we will use it for classification- if the passenger survived or passed away.

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [ ]:
training_DF = pd.read_csv(r'C:\Users\prana\OneDrive\Desktop\AI Sem 1\Appl Maths concepts for ML\Lab 2\titanic_dataset_GBC.csv')

In [ ]:
training_DF['Age'].fillna(training_DF['Age'].mean(), inplace=True)

In [ ]:
training_DF.drop('Cabin',axis=1,inplace=True)

In [ ]:
training_DF.dropna(inplace=True)

In [ ]:
sex = pd.get_dummies(training_DF['Sex'])
embark = pd.get_dummies(training_DF['Embarked'])

In [ ]:
training_DF.drop(['Sex','Embarked','Name','Ticket'],axis=1,inplace=True)

In [ ]:
training_DF = pd.concat([training_DF,sex,embark],axis=1)

In [ ]:
training_DF.drop(['female','C'],axis=1,inplace=True)

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(training_DF.drop('Survived',axis=1),
                                                    training_DF['Survived'], test_size=0.30,
                                                    random_state=42)

In [ ]:
from sklearn.linear_model import SGDClassifier
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train)

In [ ]:
# let's explore SGDClassifier parameters

'''
SGDClassifier(
    loss='hinge',
    penalty='l2',
    alpha=0.0001,
    l1_ratio=0.15,
    fit_intercept=True,
    max_iter=1000,
    tol=0.001,
    shuffle=True,
    verbose=0,
    epsilon=0.1,
    n_jobs=None,
    random_state=None,
    learning_rate='optimal',
    eta0=0.0,
    power_t=0.5,
    early_stopping=False,
    validation_fraction=0.1,
    n_iter_no_change=5,
    class_weight=None,
    warm_start=False,
    average=False,
)
'''

    # Let's form SGD models with variation in parameters loss and alpha
    # loss: 'hinge', 'log', and 'modified_huber'
    # alpha: 0.0001 and 0.001
    # explain and dicuss your findings

"\nSGDClassifier(\n    loss='hinge',\n    penalty='l2',\n    alpha=0.0001,\n    l1_ratio=0.15,\n    fit_intercept=True,\n    max_iter=1000,\n    tol=0.001,\n    shuffle=True,\n    verbose=0,\n    epsilon=0.1,\n    n_jobs=None,\n    random_state=None,\n    learning_rate='optimal',\n    eta0=0.0,\n    power_t=0.5,\n    early_stopping=False,\n    validation_fraction=0.1,\n    n_iter_no_change=5,\n    class_weight=None,\n    warm_start=False,\n    average=False,\n)\n"

In [ ]:
loss_values = ['log', 'hinge', 'modified_huber']
for loss_value in loss_values:
    model = SGDClassifier(loss=loss_value, random_state=42)
    model.fit(X_train, y_train)
    accuracy = model.score(X_test, y_test)
    print(f'Loss value: {loss_value}, Accuracy: {accuracy}')


Loss value: log, Accuracy: 0.6067415730337079
Loss value: hinge, Accuracy: 0.42696629213483145
Loss value: modified_huber, Accuracy: 0.6329588014981273


C:\Users\prana\anaconda3\lib\site-packages\sklearn\base.py:443: UserWarning: X has feature names, but SGDClassifier was fitted without feature names
  warnings.warn(
C:\Users\prana\anaconda3\lib\site-packages\sklearn\base.py:443: UserWarning: X has feature names, but SGDClassifier was fitted without feature names
  warnings.warn(
C:\Users\prana\anaconda3\lib\site-packages\sklearn\base.py:443: UserWarning: X has feature names, but SGDClassifier was fitted without feature names
  warnings.warn(


## Conclusion
Looking at the accuracy results, it seems like the 'modified_huber' loss works the best for our model, and the 'loss' function also performs quite well. However, the 'hinge' loss doesn't work as effectively and gives the lowest accuracy. This suggests that for our classification task, 'modified_huber' and 'loss' are better choices, while 'hinge' may not be the best fit.

In [ ]:
alpha_values = [0.0001, 0.001]
for alpha_value in alpha_values:
    model = SGDClassifier(alpha=alpha_value, random_state=42)
    model.fit(X_train, y_train)
    accuracy = model.score(X_test, y_test)
    print(f'Alpha value: {alpha_value}, Accuracy: {accuracy}')

Alpha value: 0.0001, Accuracy: 0.42696629213483145
Alpha value: 0.001, Accuracy: 0.6142322097378277


C:\Users\prana\anaconda3\lib\site-packages\sklearn\base.py:443: UserWarning: X has feature names, but SGDClassifier was fitted without feature names
  warnings.warn(
C:\Users\prana\anaconda3\lib\site-packages\sklearn\base.py:443: UserWarning: X has feature names, but SGDClassifier was fitted without feature names
  warnings.warn(


## Conclusion:

Loss Values:

Alpha = 0.0001: A lower alpha value (0.0001) results in a lower accuracy (0.4269). This suggests that with less regularization, the model may be overfitting to the training data, potentially capturing noise and exhibiting poorer generalization on unseen data.

Alpha = 0.001: A higher alpha value (0.001) leads to a higher accuracy (0.6142). This indicates that increasing the regularization strength may be helping the model generalize better to unseen data by preventing overfitting.